source/common/grpc/google_async_client_impl.cc

```diff
@@ -10,18 +10,62 @@
 namespace Envoy {
 namespace Grpc {
 
+GoogleAsyncClientThreadLocal::GoogleAsyncClientThreadLocal()
+    : completion_thread_(new Thread::Thread([this] { completionThread(); })) {}
+
+GoogleAsyncClientThreadLocal::~GoogleAsyncClientThreadLocal() {
+  // Force streams to shutdown and invoke TryCancel() to start the drain of
+  // pending op. If we don't do this, Shutdown() below can jam on pending ops.
+  // This is also required to satisfy the contract that once Shutdown is called,
+  // streams no longer queue any additional tags.
+  for (auto it = streams_.begin(); it != streams_.end();) {
+    (*it++)->resetStream();
+  }
+  cq_.Shutdown();
+  ENVOY_LOG(debug, "Joining completionThread");
+  completion_thread_->join();
+  ENVOY_LOG(debug, "Joined completionThread");
+  // Ensure that we have cleaned up all orphan streams, now that CQ is gone.
+  while (!streams_.empty()) {
+    (*streams_.begin())->onCompletedOps();
+  }
+}
+
+void GoogleAsyncClientThreadLocal::completionThread() {
+  ENVOY_LOG(debug, "completionThread running");
+  void* tag;
+  bool ok;
+  while (cq_.Next(&tag, &ok)) {
+    const auto& google_async_tag = *reinterpret_cast<GoogleAsyncTag*>(tag);
+    const GoogleAsyncTag::Operation op = google_async_tag.op_;
+    GoogleAsyncStreamImpl& stream = google_async_tag.stream_;
+    ENVOY_LOG(trace, "completionThread CQ event {} {}", op, ok);
+    std::unique_lock<std::mutex> lock(stream.completed_ops_lock_);
+    // It's an invariant that there must only be one pending post for arbitrary
+    // length completed_ops_, otherwise we can race in stream destruction, where
+    // we process multiple events in onCompletedOps() but have only partially
+    // consumed the posts on the dispatcher.
```

Comment:
The only problem I see with this approach is that it isn't compatible with limiting the number of completed ops processed in the dispatcher post before relinquishing control back to the dispatcher, which could lead to starving the silo thread from processing other events if there are a huge number of completed ops. I think that's OK for now.

Reply:
Yeah, fair point. I've added a TODO, we can be smarter about bounding completion run sizes in the future if this does manifest as an issue.